src/test/pages/EventDetails.test.tsx

```diff
@@ -0,0 +1,230 @@
+    expect(image).toHaveAttribute("src", "http://placehold.it/500x276");
+  });
+
+  it("shows a map and address details if location, latitude and longitude are specified, and event is not tagged as online", async () => {
+    const mapLocation = {
+      address: { addressLine1: "Mock street", town: "Mock town", postalCode: "ABC 123" },
+      latitude: 53.95038292218263,
+      longitude: -1.0510007751299424,
+    };
+    await setupTest({
+      role: "STUDENT",
+      event: { ...mockEvent, tags: ["booster", "student"], location: mapLocation } as IsaacEventPageDTO,
+    });
+    const { map, location } = getFields();
+    expect(map).toBeInTheDocument();
+    expect(location).toHaveTextContent(formatAddress(mapLocation));
+  });
+
+  it("shows location as online if event is tagged as virtual", async () => {
+    await setupTest({ role: "STUDENT", event: { ...mockEvent, tags: ["virtual"] } as IsaacEventPageDTO });
+    const { location } = getFields();
+    expect(location).toHaveTextContent("Online");
+  });
+
+  it("shows private event badge if event is private", async () => {
+    await setupTest({ role: "STUDENT", event: { ...mockEvent, privateEvent: true } as IsaacEventPageDTO });
+    const { privateBadge } = getFields();
+    expect(privateBadge).toBeVisible();
+  });
+
+  it("does not show private event badge if even is not private", async () => {
+    await setupTest({ role: "STUDENT", event: { ...mockEvent, privateEvent: false } as IsaacEventPageDTO });
+    const { privateBadge } = getFields();
+    expect(privateBadge).not.toBeInTheDocument();
+  });
+
+  it.each(["ADMIN", "EVENT_MANAGER", "CONTENT_EDITOR"] as TestUserRole[])(
+    "if user is %s, a google calendar button shows and can be clicked",
+    async (role) => {
+      const mockGoogleCalendarTemplate = jest.fn();
+      jest.spyOn(eventServices, "googleCalendarTemplate").mockImplementation(mockGoogleCalendarTemplate);
+      await setupTest({ role: role, event: mockEvent as IsaacEventPageDTO });
+      const { googleCalendarButton } = getFields();
+      expect(googleCalendarButton).toBeVisible();
+      await userEvent.click(googleCalendarButton as HTMLElement);
+      expect(mockGoogleCalendarTemplate).toHaveBeenCalled();
+    },
+  );
+
+  it("if user is STUDENT, a google calendar button does not show", async () => {
+    await setupTest({ role: "STUDENT", event: mockEvent as IsaacEventPageDTO });
+    const { googleCalendarButton } = getFields();
+    expect(googleCalendarButton).not.toBeInTheDocument();
+  });
+
+  it("shows event date and time, and message if event was in the past", async () => {
+    const startDate = new Date(new Date().setMonth(new Date().getMonth() - 2));
+    const endDate = new Date(startDate.getTime() + 2 * 60 * 60 * 1000);
+    const pastEvent = { ...mockEvent, date: startDate, endDate: endDate } as IsaacEventPageDTO;
+    await setupTest({
+      role: "STUDENT",
+      event: pastEvent as IsaacEventPageDTO,
```

Comment:
Unnecessary casting. To be honest, the fact that we have so many castings is a code smell.

Reply:
This was an odd one; I'd done a lot of castings in the tests because in the data file it didn't like accepting the mockEvent as an IsaacEventPageDTO (even though I'd structured it by taking a real API response and editing it!).  It seems like there were some typing errors in the first place so I made some changes which allowed me to remove most of the casting.